src/Draco.Compiler/Syntax/ParseTreeVisitor.cs

```diff
@@ -0,0 +1,147 @@
+using System;
+using static Draco.Compiler.Syntax.ParseTree;
+
+namespace Draco.Compiler.Syntax;
+
+/// <summary>
+/// Defines a visitor of a <see cref="ParseTree"/>.
+/// </summary>
+/// <typeparam name="T">The return type of the visitor.</typeparam>
+internal interface IParseTreeVisitor<out T>
+{
+    T VisitCompilationUnit(CompilationUnit compilationUnit);
+
+    T VisitDecl(Decl decl);
+    T VisitFuncDecl(Decl.Func decl);
+    T VisitLabelDecl(Decl.Label decl);
+    T VisitVariableDecl(Decl.Variable decl);
+
+    T VisitFuncParam(FuncParam param);
+
+    T VisitFuncBody(FuncBody body);
+    T VisitBlockFuncBody(FuncBody.BlockBody body);
+    T VisitInlineFuncBody(FuncBody.InlineBody body);
+
+    T VisitTypeExpr(TypeExpr typeExpr);
+    T VisitNameTypeExpr(TypeExpr.Name typeExpr);
+
+    T VisitTypeSpecifier(TypeSpecifier specifier);
+
+    T VisitStmt(Stmt stmt);
+    T VisitDeclStmt(Stmt.Decl stmt);
+    T VisitExprStmt(Stmt.Expr stmt);
+
+    T VisitExpr(Expr expr);
+    T VisitBlockExpr(Expr.Block expr);
+    T VisitIfExpr(Expr.If expr);
+    T VisitWhileExpr(Expr.While expr);
+    T VisitGotoExpr(Expr.Goto expr);
+    T VisitReturnExpr(Expr.Return expr);
+    T VisitLiteralExpr(Expr.Literal expr);
+    T VisitFuncCallExpr(Expr.FuncCall expr);
+    T VisitIndexExpr(Expr.Index expr);
+    T VisitVariableExpr(Expr.Variable expr);
+    T VisitMemberAccessExpr(Expr.MemberAccess expr);
+    T VisitUnaryExpr(Expr.Unary expr);
+    T VisitBinaryExpr(Expr.Binary expr);
+    T VisitGroupingExpr(Expr.Grouping expr);
+}
+
+/// <summary>
+/// Provides a base implementation of <see cref="IParseTreeVisitor{T}"/>.
+/// </summary>
+/// <typeparam name="T">The return type of the visitor.</typeparam>
+internal abstract class BaseParseTreeVisitor<T> : IParseTreeVisitor<T>
```

Comment:
Isn't a Base suffix the convention? So `ParseTreeVisitorBase`? I'm honestly not sure